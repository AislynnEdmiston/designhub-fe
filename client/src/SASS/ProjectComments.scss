@import './palette';
@import './mixins';

.project-comments {
  width: 31%;
  height: 100%;
  background-color: #15161a;
  border: 5px solid #15161a;
  box-sizing: border-box;
  border-radius: 10px;
  padding: 14px;
  display: flex;
  flex-flow: column nowrap;
  margin: 0 0 0 1%;
  @include themify($themes) {
    background: themed('bg2');
    border: 5px solid themed('bg2');
  }

  .comments-header {
    letter-spacing: 0.11em;
    font-weight: normal;
    font-size: 1rem;
    margin-bottom: 20px;
  }

  .comments-body {
    width: 100%;
    height: 100%;
    overflow-y: scroll;
    font-size: 0.9rem;
    line-height: 20px;
    display: flex;
    flex-direction: column;
    align-items: flex-end;

    .ProjectComment__body {
      align-items: center;
      background: $blue;
      border-radius: 10px;
      box-sizing: border-box;
      display: flex;
      flex-direction: row-reverse;
      margin-top: 20px;
      padding: 10px 8px;
      width: max-content;
      max-width: 100%;
      height: auto;
      &.--you {
        color: white;
      }

      &.--them {
        background: none;
        flex-direction: row;
        width: 100%;
        .ProjectComment__body__text {
          text-align: left;
        }
      }

      .ProjectComment__body__avatar {
        align-self: baseline;
        border-radius: 100%;
        height: 35px;
        object-fit: cover;
        width: 35px;
      }

      .ProjectComment__body__text {
        font-size: 12px;
        line-height: 14px;
        letter-spacing: 0.11em;
        padding: 0 8px;
        width: max-content;
        max-width: 250px;
        overflow-wrap: break-word;
        text-align: right;

        header {
          display: flex;
        }
        p {
          font-weight: normal;
          margin-top: 2px;
          height: auto;
        }
        .username {
          margin-bottom: 10px;
        }
      }
    }
  }

  .comments-form {
    padding-top: 20px;
    .form-wrapper {
      align-items: center;
      // border: 1px solid red;
      background: #2a2b31;
      border-radius: 10px;
      display: flex;
      justify-content: space-between;
      padding: 2%;
      @include themify($themes) {
        background: themed('bg');
      }
    }
    input[type='text'] {
      border: none;
      background: #2a2b31;
      border-radius: 10px;
      box-sizing: border-box;
      color: white;
      font-size: 0.8rem;
      font-weight: 400;
      outline: none;
      width: 85%;
      padding: 2%;
      overflow-wrap: break-word;
      max-height: 200px;
      @include themify($themes) {
        background: themed('bg');
        color: themed('text');
      }
    }
    button {
      background-color: #5557fe;
      border: none;
      border-radius: 100%;
      cursor: pointer;
      height: 40px;
      margin-bottom: 2%;
      outline: none;
      width: 40px;
      z-index: 9;
      svg {
        margin: 0 0 -3px -3px;
        text-align: center;
        vertical-align: middle;
      }
      &:hover {
        background: $btn-primary-hover;
      }
    }
  }
}
