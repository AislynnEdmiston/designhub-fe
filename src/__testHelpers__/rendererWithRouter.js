import React from 'react'
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux'
import { createStore, applyMiddleware } from 'redux'
import { createMemoryHistory } from 'history';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';

import rootReducer from '../store/reducers';
jest.mock('../auth-wrapper');

const store = createStore(
  rootReducer,
  applyMiddleware(thunk)
);
const rendererWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) => {
  return {
    ...renderer.create(
      <Provider store={store}>
        <Router history={history}>
          {ui}
        </Router>
      </Provider>
    ),
    history
  }
}

export default rendererWithRouter;
