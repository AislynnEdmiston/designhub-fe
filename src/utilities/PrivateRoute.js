import React from 'react';
import { Route } from 'react-router-dom';
import { useAuth0 } from './auth-wrapper.js';
import LandingPage from '../views/LandingPage/LandingPage.js';

const PrivateRoute = ({ component: Component, path, ...rest }) => {
  const { isAuthenticated } = useAuth0();

  const render = (props) =>
    isAuthenticated === true ? <Component {...props} /> : <LandingPage />;

  return <Route path={path} render={render} {...rest} />;
};

export default PrivateRoute;
