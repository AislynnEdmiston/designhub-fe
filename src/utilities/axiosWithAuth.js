import axios from 'axios';

export const axiosWithAuth = () => {
  const token = localStorage.getItem('token');

  return axios.create({
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`
    },
    baseURL: `${process.env.REACT_APP_GQL_API}`
  });
};
