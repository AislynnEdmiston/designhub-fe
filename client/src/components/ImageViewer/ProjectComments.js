import React, { useState } from 'react';
import axios from 'axios';
import moment from 'moment';
import SendIcon from '../Icons/SendIcon';
import '../../SASS/ProjectComments.scss';

const ProjectComments = ({
  activeUser,
  addComments,
  comments,
  thisProject
}) => {
  // console.log('ProjectComments.js RENDER comments', comments);
  // console.log('ProjectComments.js RENDER activeUser', activeUser);

  // ref for bottom of comments feed
  const [commentAnchor, setCommentAnchor] = useState(null);
  //function to autoscroll to commentAnchor
  const scrollToBottom = () => {
    commentAnchor.scrollIntoView({ behavior: 'smooth' });
  };

  //local state for form input
  const [newComment, setNewComment] = useState('');

  //click submit
  const handleSubmit = async e => {
    e.preventDefault();
    if (!newComment) return; //dont submit blank comments

    const thisComment = {
      userId: activeUser.id,
      username: activeUser.username,
      projectId: thisProject.id,
      text: newComment
    };
    // console.log('ProjectComments.js handleSubmit() thisCOmment', thisComment);

    try {
      const res = await axios.post(
        `${process.env.REACT_APP_BASE_URL}api/v1/comments/project`,
        thisComment
      );
      const newComment = res.data.data[0];

      //glue the avatar back on and insert into local state so we don't have to reload the component
      newComment.userAvatar = activeUser.avatar;
      const updateComments = [...comments, newComment];
      addComments(updateComments);
      setNewComment('');
    } catch (err) {
      console.log('ProjectComments.js handleSubmit() ERROR', err);
    }
  };

  return (
    <div className="project-comments">
      <header className="comments-header">Comments</header>

      <section className="comments-body">
        {commentAnchor && scrollToBottom()}

        {comments.map(c => (
          <div className="ProjectComment__wrapper">
            <section
              className={
                activeUser.id === c.userId
                  ? 'ProjectComment__body --you'
                  : 'ProjectComment__body --them'
              }
            >
              {activeUser.id === c.userId ? null : (
                <img
                  src={c.userAvatar}
                  alt="avatar"
                  className="ProjectComment__body__avatar"
                />
              )}
              <div className="ProjectComment__body__text">
                <header>
                  <em>{activeUser.id === c.userId ? 'You' : c.username}</em>
                  {/* <p>{moment(c.created_at).fromNow()}</p> */}
                </header>
                <p>{c.text}</p>
              </div>
            </section>
          </div>
        ))}

        <div ref={el => setCommentAnchor(el)}></div>
      </section>

      <section className="comments-form">
        <form onSubmit={handleSubmit}>
          <div className="form-wrapper">
            <input
              type="text"
              value={newComment}
              onChange={e => setNewComment(e.target.value)}
              placeholder="Leave a comment..."
            />
            <button>
              <SendIcon />
            </button>
          </div>
        </form>
      </section>
    </div>
  );
};

export default ProjectComments;
