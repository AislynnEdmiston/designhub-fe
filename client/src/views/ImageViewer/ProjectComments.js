import React, { useState } from 'react';
import { axiosWithAuth } from '../../utilities/axiosWithAuth.js';
import { connect } from 'react-redux';
import { useWindowDimensions } from './useWindowDimensions.js';
import { Link } from 'react-router-dom';
import SendIcon from '../../common/Icons/SendIcon';

import { addProjectComment } from '../../store/actions';

import './SASS/ProjectComments.scss';

const ProjectComments = ({
  activeUser,
  comments,
  modal,
  thisProject,
  addProjectComment
}) => {
  //custom hook to get window height/width
  const { width } = useWindowDimensions();
  // ref for bottom of comments feed
  const [commentAnchor, setCommentAnchor] = useState(null);
  //function to autoscroll to commentAnchor
  const scrollToBottom = () => {
    //scroll comments window ONLY at desktop widths
    if (width <= 1024) return;
    else commentAnchor.scrollIntoView({ behavior: 'smooth' });
  };

  //local state for form input
  const [newComment, setNewComment] = useState('');

  //function for sending comment notifications

  const postCommentNotification = async (
    username,
    commentText,
    projectId,
    invitedUserId,
    activeUserId,
    mainImgUrl,
    commentsId,
    activeUserAvatar,
    type
  ) => {
    axiosWithAuth().post('api/v1/invite/comments', {
      activeUsername: username,
      commentText: commentText,
      projectId: projectId,
      invitedUserId: invitedUserId,
      activeUserId: activeUserId,
      mainImgUrl: mainImgUrl,
      commentsId: commentsId,
      activeUserAvatar: activeUserAvatar,
      type: type
    });
  };

  //click submit
  const handleSubmit = async e => {
    e.preventDefault();
    if (!newComment) return; //dont submit blank comments

    const thisComment = {
      userId: activeUser.id,
      username: activeUser.username,
      projectId: thisProject.id,
      text: newComment
    };

    try {
      const res = await addProjectComment(thisComment);
      const newComment = res.data.data[0];

      if (thisProject.id !== activeUser.id) {
        await postCommentNotification(
          activeUser.username,
          newComment.text,
          thisProject.id,
          thisProject.userId,
          activeUser.id,
          thisProject.mainImg,
          newComment.id,
          activeUser.avatar,
          'comment'
        );
      }

      //glue the avatar back on and insert into local state so we don't have to reload the component
      setNewComment('');
    } catch (err) {
      console.log('ProjectComments.js handleSubmit() ERROR', err);
    }
  };

  return (
    <div className="project-comments">
      <header className="comments-header">Comments</header>
      <section className="comments-body">
        {comments &&
          comments.map(c => (
            <div
              key={c.id}
              className={
                activeUser.id === c.userId
                  ? 'ProjectComment__body --you'
                  : 'ProjectComment__body --them'
              }
            >
              {activeUser.id === c.userId ? null : (
                <Link to={`/profile/${c.userId}/${c.username}`}>
                  <img
                    src={c.userAvatar}
                    alt="avatar"
                    className="ProjectComment__body__avatar"
                  />
                </Link>
              )}
              <div className="ProjectComment__body__text">
                <Link
                  to={`/profile/${c.userId}/${c.username}`}
                  className="user-links"
                >
                  <p className="username">
                    {activeUser.id === c.userId ? `${c.username}` : c.username}
                  </p>
                </Link>
                <p>{c.text}</p>
              </div>
            </div>
          ))}
        {/* this ref grabs the bottom of the comments display and scrolls it into view, automatically disabled at the column-view breakpoint */}
        <div ref={el => setCommentAnchor(el)}></div>
        {commentAnchor && !modal && scrollToBottom()}
      </section>

      <section className="comments-form">
        <form onSubmit={handleSubmit}>
          <div className="form-wrapper">
            <input
              type="text"
              value={newComment}
              onChange={e => setNewComment(e.target.value)}
              placeholder="Leave a comment..."
            />
            <button>
              <SendIcon />
            </button>
          </div>
        </form>
      </section>
    </div>
  );
};

const mapStateToProps = state => {
  return {
    comments: state.comments.projectComments
  };
};

export default connect(
  mapStateToProps,
  { addProjectComment }
)(ProjectComments);
