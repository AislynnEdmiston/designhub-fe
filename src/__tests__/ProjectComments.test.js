import React from 'react';
import ProjectComments from '../views/ImageViewer/ProjectComments';
import { useAuth0 } from '../utilities/auth-wrapper';
import rendererWithRouter from '../__testHelpers__/rendererWithRouter';

const comment = {
  username: 'johndoe@me.com',
  commentText: 'testing',
  projectId: 1,
  invitedUserId: 2,
  activeUserId: 1,
  commentsId: 3,
};

jest.mock('../utilities/auth-wrapper');
describe('<ProjectComments />', () => {
  beforeEach(() => {
    useAuth0.mockReturnValue({
      comment,
      postCommentNotification: jest.fn(),
    });
  });

  describe('should match the previous snapshot of the component', () => {
    it('tests that the component has not changed and matched the previous snapshot', () => {
      const tree = rendererWithRouter(
        <ProjectComments history={{ location: { pathname: '/project' } }} />
      );
      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
