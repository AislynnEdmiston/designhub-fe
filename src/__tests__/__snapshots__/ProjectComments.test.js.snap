// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectComments /> should match the previous snapshot of the component tests that the component has not changed and matched the previous snapshot 1`] = `
<div
  className="project-comments"
>
  <header
    className="comments-header"
  >
    Comments
  </header>
  <section
    className="comments-body"
  >
    <div />
  </section>
  <section
    className="comments-form"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-wrapper"
      >
        <input
          onChange={[Function]}
          placeholder="Leave a comment..."
          type="text"
          value=""
        />
        <button>
          <svg
            fill="none"
            height="19"
            viewBox="0 0 19 19"
            width="19"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17.4168 1.5835L8.7085 10.2918"
              stroke="white"
              strokeLinecap="round"
              strokeLinejoin="round"
            />
            <path
              d="M17.4168 1.5835L11.8752 17.4168L8.7085 10.2918L1.5835 7.12516L17.4168 1.5835Z"
              stroke="white"
              strokeLinecap="round"
              strokeLinejoin="round"
            />
          </svg>
        </button>
      </div>
    </form>
  </section>
</div>
`;
